### ISSUE TITLE
Paper candidates: pattern mining

### ISSUE THREAD

**Issue by bzz** [src-d/reading-club#43]
One of the things that might be a big topic for src-d this year and thus was suggested by @vmarkovtsev was to start looking into the papers regarding pattern mining in source code.

This issue is to be an ☂️to list such papers out, even without attachment to any reading session.

**Comment by Jan21** [comment]
[Getafix: Learning to fix bugs automatically (2019)](https://arxiv.org/pdf/1902.06111v1.pdf)

> Static analyzers, including linters, can warn developers about programming errors, bugs, style errors, and generally suspicious code. Sometimes analyzer rules have obvious fixes but other times choosing an appropriate fix is more nuanced and can take some time for a developer to proceed. Offering fix suggestions for these trickier analyzers can therefore help save developer time, but in order for developers to accept the offered fix the patch must look natural. Additionally, in a production environment there is a limited amount of resources that can be spent validating fix candidates, so there is not time to evaluate a large number of fix candidates. In fact, at Facebook we can generally only afford to validate a single suggested fix. To solve these issues, we built Getafix, a tool which mines patterns from past fixes made by developers with enough context to accurately create natural-looking autofix suggestions. The patches we learn contain all the context needed to apply and rank them, so even though we only validate the top fix candidate we still are able to offer fix suggestions in a lot of cases. Additionally, our pattern mining approach can be directed at the whole code change history to find new lint warnings, avoiding the manual effort from developers to create new lint rules. Our pattern mining approach is more effective in discovering new and applicable patterns than previous work because it includes context about where and how the pattern should apply.

This paper describes interesting idea for pattern mining based on Anti-unification which is technique from inductive logic programming. They say they are already using it in production at Facebook

**Comment by Jan21** [comment]
[Learning Quick Fixes from Code Repositories (2018)](https://arxiv.org/pdf/1803.03806.pdf)

> Code analyzers such as Error Prone and FindBugs
detect code patterns symptomatic of bugs, performance issues,
or bad style. These tools express patterns as quick fixes that
detect and rewrite unwanted code. However, it is difficult to
come up with new quick fixes and decide which ones are useful
and frequently appear in real code. We propose to rely on the
collective wisdom of programmers and learn quick fixes from
revision histories in software repositories. We present REVISAR,
a tool for discovering common Java edit patterns in code
repositories. Given code repositories and their revision histories,
REVISAR (i) identifies code edits from revisions and (ii) clusters
edits into sets that can be described using an edit pattern. The
designers of code analyzers can then inspect the patterns and add
the corresponding quick fixes to their tools. We ran REVISAR on
nine popular GitHub projects, and it discovered 89 useful edit
patterns that appeared in 3 or more projects. Moreover, 64% of
the discovered patterns did not appear in existing tools. We then
conducted a survey with 164 programmers from 124 projects
and found that programmers significantly preferred eight out of
the nine of the discovered patterns. Finally, we submitted 16 pull
requests applying our patterns to 9 projects and, at the time of
the writing, programmers accepted 6 (60%) of them. The results
of this work aid toolsmiths in discovering quick fixes and making
informed decisions about which quick fixes to prioritize based
on patterns programmers actually apply in practice.
Index Terms—code repositories, mining, program analysis
tools, program transformation, quick fixes

Similar to the Getafix, I think Getafix took an inspiration from this paper.

**Comment by bzz** [comment]
[Aroma: Code Recommendation via Structural Code Search (2018)](https://arxiv.org/abs/1812.01158)

>Programmers often write code which have similarity to existing code written somewhere. A tool that could help programmers to search such similar code would be immensely useful. Such a tool could help programmers to extend partially written code snippets to completely implement necessary functionality, help to discover extensions to the partial code which are commonly done by other programmers, help to cross-check against similar code written by other programmers, or help to add extra code which would avoid common mistakes and errors. We propose Aroma, a tool and technique for code recommendation via structural code search. Aroma indexes a huge code corpus including thousands of open-source projects, takes a partial code snippet as input, searches the indexed method bodies which contain the partial code snippet, clusters and intersects the results of search to recommend a small set of succinct code snippets which contain the query snippet and which appears as part of several programs in the corpus. We evaluated Aroma on several randomly selected queries created from the corpus and as well as those derived from the code snippets obtained from Stack Overflow, a popular website for discussing code. We found that Aroma was able to retrieve and recommend most relevant code snippets efficiently.

Another publication by FB, after Getafix.

**Comment by bzz** [comment]
[On the Feasibility of Transfer-learning Code Smells using Deep Learning 2019](https://arxiv.org/abs/1904.03031)
> Context: A substantial amount of work has been done to detect smells in source code using metrics-based and heuristics-based methods. Machine learning methods have been recently applied to detect source code smells; however, the current practices are considered far from mature. Objective: First, explore the feasibility of applying deep learning models to detect smells without extensive feature engineering, just by feeding the source code in tokenized form. Second, investigate the possibility of applying transfer-learning in the context of deep learning models for smell detection. Method: We use existing metric-based state-of-the-art methods for detecting three implementation smells and one design smell in C# code. Using these results as the annotated gold standard, we train smell detection models on three different deep learning architectures. These architectures use Convolution Neural Networks (CNNs) of one or two dimensions, or Recurrent Neural Networks (RNNs) as their principal hidden layers. For the first objective of our study, we perform training and evaluation on C# samples, whereas for the second objective, we train the models from C# code and evaluate the models over Java code samples. We perform the experiments with various combinations of hyper-parameters for each model. Results: We find it feasible to detect smells using deep learning methods. Our comparative experiments find that there is no clearly superior method between CNN-1D and CNN-2D. We also observe that performance of the deep learning models is smell-specific. Our transfer-learning experiments show that transfer-learning is definitely feasible for implementation smells with performance comparable to that of direct-learning. This work opens up a new paradigm to detect code smells by transfer-learning especially for the programming languages where the comprehensive code smell detection tools are not available.

**Comment by bzz** [comment]
Closing, as this seem to never pick up any internal or external traction. Feel free to re-open later!
